proxy/grpcproxy/cluster.go

```diff
@@ -15,38 +15,177 @@
+	gr   *naming.GRPCResolver
+
+	// self is true to return its own client listen URLs
+	// for member list API
+	self  bool
+	laddr string
+
+	wa   gnaming.Watcher
+	umu  sync.RWMutex
+	umap map[string]struct{}
+}
+
+// NewClusterProxy takes optional prefix to fetch grpc-proxy member endpoints.
+// 'self' is true then member list API returns the client listen addresses.
+// The returned channel is closed when there is grpc-proxy endpoint registered
+// and the client's context is canceled so the 'register' loop returns.
+func NewClusterProxy(c *clientv3.Client, self bool, laddr string, prefix string) (pb.ClusterServer, <-chan struct{}, error) {
+	if self && prefix != "" {
+		return nil, nil, fmt.Errorf("both 'self' and 'prefix' are configured")
+	}
+	donec := make(chan struct{})
+
+	cp := &clusterProxy{
+		clus: c.Cluster,
+		ctx:  c.Ctx(),
+		gr:   &naming.GRPCResolver{Client: c},
+
+		self:  self,
+		laddr: laddr,
+	}
+	if prefix == "" {
+		close(donec)
+		return cp, donec, nil
+	}
+
+	go cp.resolve(prefix, donec)
+	return cp, donec, nil
+}
+
+// allow maximum 1 retry per second
+const resolveRetryRate = 1
+
+func (cp *clusterProxy) resolve(prefix string, donec chan struct{}) {
+	rm := rate.NewLimiter(rate.Limit(resolveRetryRate), resolveRetryRate)
+	for rm.Wait(cp.ctx) == nil {
+		watcher, err := cp.gr.Resolve(prefix)
+		if err != nil {
+			plog.Warningf("failed to resolve %q (%v)", prefix, err)
+			continue
+		}
+		cp.wa = watcher
+		break
+	}
+
+	// pass 'donec' to signal monitor routine return
+	go cp.monitor(donec)
 }
 
-func NewClusterProxy(c *clientv3.Client) pb.ClusterServer {
-	return &clusterProxy{
-		client: c,
+func (cp *clusterProxy) monitor(donec chan struct{}) {
+	// to signal monitor routine return
+	defer close(donec)
+
+	for cp.ctx.Err() == nil {
+		ups, err := cp.wa.Next()
+		if err != nil {
+			plog.Warningf("clusterProxy watcher error (%v)", err)
+			if grpc.ErrorDesc(err) == naming.ErrWatcherClosed.Error() {
+				return
+			}
+		}
+
+		cp.umu.Lock()
+		if len(cp.umap) == 0 {
+			cp.umap = make(map[string]struct{}, len(ups))
+		}
+		for i := range ups {
+			switch ups[i].Op {
+			case gnaming.Add:
+				cp.umap[ups[i].Addr] = struct{}{}
+			case gnaming.Delete:
+				delete(cp.umap, ups[i].Addr)
+			}
+		}
+		cp.umu.Unlock()
+	}
+}
+
+func (cp *clusterProxy) members() []*pb.Member {
+	cp.umu.RLock()
+	defer cp.umu.RUnlock()
+
+	// using its own listen client URLs
+	// or addresses haven't been populated from the watcher
+	if cp.self || len(cp.umap) == 0 {
+		return []*pb.Member{{ClientURLs: []string{cp.laddr}}}
+	}
+
+	mbs := make([]*pb.Member, 0, len(cp.umap))
+	for addr := range cp.umap {
+		mbs = append(mbs, &pb.Member{ClientURLs: []string{addr}})
 	}
+	return mbs
 }
 
 func (cp *clusterProxy) MemberAdd(ctx context.Context, r *pb.MemberAddRequest) (*pb.MemberAddResponse, error) {
-	conn := cp.client.ActiveConnection()
-	return pb.NewClusterClient(conn).MemberAdd(ctx, r)
+	var ps []string
+	if r != nil {
```

Comment:
is this error check necessary? I think panicking on nil for cp.clus.MemberRemove(ctx, r.PeerURLs) is OK